### PR TITLE
safekeeper: add walreceiver metrics

### LINKED ISSUE #9328: safekeeper: add metric for per-timeline queue depth/buffer size
When shipping WAL from safekeepers to pageservers, we presumably buffer the WAL records. We should have a metric tracking how much WAL is buffered or in-flight.

Ideally per timeline, but this has very high cardinality. Optionally, we can consider a histogram.

Also investigate current buffer sizes and limits.

### PR DESCRIPTION
**Pull request by erikgrinaker**
## Problem

We don't have any observability for Safekeeper WAL receiver queues.

Resolves #9328.

## Summary of changes

Adds a few WAL receiver metrics:

* `safekeeper_wal_receivers`: gauge of currently connected WAL receivers.
* `safekeeper_wal_receiver_queue_depth`: histogram of queue depths per receiver, sampled every 5 seconds.
* `safekeeper_wal_receiver_queue_depth_total`: gauge of total queued messages across all receivers.
* `safekeeper_wal_receiver_queue_size_total`: gauge of total queued message sizes across all receivers.

There are already metrics for ingested WAL volume: `written_wal_bytes` counter per timeline, and `safekeeper_write_wal_bytes` per-request histogram.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by erikgrinaker** [comment]
@VladLazar @jcsp Draft of WAL receiver queue depth metric, for an early look. This also rewrites the WAL receiver loop to be event driven, let's split that out to a separate PR if we agree on the approach.

I went for a histogram rather than a high-cardinality metric, since I'm not sure we care about individual timelines (and we can add e.g. logging or other observability if we really need to know which timelines are backpressuring).

Any other metrics we want to add? Might be useful with the byte size of buffered messages -- in aggregate we can easily get this by having counters for WAL bytes received and WAL bytes processed, and compute the difference to find the buffer volume. If we want it per queue we'll need to wrap the `mpsc` queue to track sizes on `send()` and `recv()`.

**Review by VladLazar — COMMENTED** [review]
This makes sense to me.

---

> Any other metrics we want to add?

Would it be interesting to know the worst offender timelines? Perhaps not since we know it's bounded to approx 40 MiB from your analysis [here](https://github.com/neondatabase/neon/issues/9328#issuecomment-2417062276).

> In addition to this, it would also be interesting to know how much WAL we're keeping around on local disk until the Pageserver has flushed it to the S3 LSM.

I think we already know this indirectly, but a more direct metric would be nice:
```
commit_lsn - max(min_remote_consistent_lsn, backup_lsn)
where:
min_remote_consistent_lsn = min remote consistent LSN across all shard for the tenant+timeline in question
backup_lsn = latest LSN uploaded to SK to remote storage
commit_lsn = latest LSN written to disk by sk
```

File: safekeeper/src/receive_wal.rs [hunk 1]
```diff
@@ -493,67 +494,78 @@ impl WalAcceptor {
...
-            let reply_msg = if matches!(next_msg, ProposerAcceptorMessage::AppendRequest(_)) {
-                // Loop through AppendRequests while available to write as many WAL records as
-                // possible without fsyncing.
-                //
-                // Make sure the WAL is flushed before returning, see:
-                // https://github.com/neondatabase/neon/issues/9259
-                //
-                // Note: this will need to be rewritten if we want to read non-AppendRequest messages here.
-                // Otherwise, we might end up in a situation where we read a message, but don't
-                // process it.
```
[the remark this answers is not in the document]
Comment (erikgrinaker) [reply safekeeper/src/receive_wal.rs:516]: No, this rewrite addresses the hazard.

**Comment by erikgrinaker** [comment]
> Would it be interesting to know the worst offender timelines?

As a metric, it would be high-cardinality. I suggest we instead e.g. log when a send has been blocked for n seconds, and we can get the timelines that way.

> I think we already know this indirectly, but a more direct metric would be nice

Yeah, should be trivial to compute.

**Comment by VladLazar** [comment]
> > Would it be interesting to know the worst offender timelines?
> 
> As a metric, it would be high-cardinality. I suggest we instead e.g. log when a send has been blocked for n seconds, and we can get the timelines that way.
> 
> > I think we already know this indirectly, but a more direct metric would be nice
> 
> Yeah, should be trivial to compute.

> As a metric, it would be high-cardinality. I suggest we instead e.g. log when a send has been blocked for n seconds, and we can get the timelines that way.

I was thinking of a metric with a limited number of slots where we register and deregister offending timelines as we go. Logging is easier though.

**Comment by erikgrinaker** [comment]
Yeah, but we'd have to include the timeline ID as a dimension (right?), and over time different timelines would be in the top k, increasing the cardinality to a large number of timelines.

**Comment by VladLazar** [comment]
> Yeah, but we'd have to include the timeline ID as a dimension (right?), and over time different timelines would be in the top k, increasing the cardinality to a large number of timelines.

Hmm, yeah I guess that still counts as cardinality.

**Comment by erikgrinaker** [comment]
I split the WAL acceptor refactor out to #9462.

File: safekeeper/src/metrics.rs [hunk 2]
```diff
@@ -204,6 +204,28 @@ pub static WAL_BACKUP_TASKS: Lazy<IntCounterPair> = Lazy::new(|| {
...
+        "Number of currently connected WAL receivers (i.e. connected computes)"
+    )
+    .expect("Failed to register safekeeper_wal_receivers")
+});
+pub static WAL_RECEIVER_QUEUE_DEPTH: Lazy<Histogram> = Lazy::new(|| {
+    // Use powers of two buckets, but add a bucket at 0 and the max queue size to track empty and
+    // full queues respectively.
+    let mut buckets = pow2_buckets(1, MSG_QUEUE_SIZE);
+    buckets.insert(0, 0.0);
+    buckets.insert(buckets.len() - 1, (MSG_QUEUE_SIZE - 1) as f64);
```
Comment (erikgrinaker) [inline safekeeper/src/metrics.rs:219]: A bit unsure about the buckets here.

Powers of two gives poor resolution at the top -- we could consider e.g. linear buckets of 32, either from 0 or powers-of-two up to 64 then linear. But it's probably fine.

Inserting a bucket at 255 is also perhaps a bit odd. The motivation was to get a separate bucket for full queues. We could also consider shifting all buckets down by 1 (so 0,1,3,7,15,...255,+Inf) and/or using infinity for full queues.

File: safekeeper/src/metrics.rs [hunk 3]
```diff
@@ -204,6 +204,28 @@ pub static WAL_BACKUP_TASKS: Lazy<IntCounterPair> = Lazy::new(|| {
     )
     .expect("Failed to register safekeeper_wal_backup_tasks_finished_total counter")
 });
+pub static WAL_RECEIVERS: Lazy<IntGauge> = Lazy::new(|| {
+    register_int_gauge!(
+        "safekeeper_wal_receivers",
```
Comment (erikgrinaker) [inline safekeeper/src/metrics.rs:209]: We can also get this from summing up `safekeeper_connected_computes` across all timelines, but seemed useful to have a total gauge. Particularly for comparison with the number of samples from `safekeeper_wal_receiver_queue_depth`.

File: safekeeper/src/receive_wal.rs [hunk 4]
```diff
@@ -445,6 +448,12 @@ async fn network_write<IO: AsyncRead + AsyncWrite + Unpin>(
 /// walproposer, even when it's writing a steady stream of messages.
 const FLUSH_INTERVAL: Duration = Duration::from_secs(1);
 
+/// The metrics computation interval.
+///
+/// The Prometheus poll interval is 60 seconds at the time of writing. We sample the queue depth
+/// every 5 seconds, for 12 samples per poll. This will give a count of up to 12x active timelines.
+const METRICS_INTERVAL: Duration = Duration::from_secs(5);
```
Comment (erikgrinaker) [inline safekeeper/src/receive_wal.rs:455]: We could also just do a single sample per timeline per poll interval (every 60 seconds), such that the number of observations matches the number of timelines, but I think we probably want more frequent sampling.

**Review by erikgrinaker — COMMENTED** [review]
This should be good to go now. I'll submit a separate PR for logging of blocked timelines (and probably roundtrip append latencies if we don't have them already).

>> I think we already know this indirectly, but a more direct metric would be nice
>
>Yeah, should be trivial to compute.

Ah, these are per-timeline Prometheus metrics. Let's not add another computed metric, since these are per-timeline (thus expensive) -- we can compute them in a dashboard query.

File: safekeeper/src/metrics.rs (hunk 2 above)
Comment (VladLazar) [reply safekeeper/src/metrics.rs:219]: I think this is fine. We can tweak as we go.

**Review by VladLazar — APPROVED** [review]
Looks good. I was a bit surprised to see that the queue depth metrics tracks messages and not bytes. Do you have thoughts on this?

File: safekeeper/src/receive_wal.rs (hunk 4 above)
Comment (VladLazar) [reply safekeeper/src/receive_wal.rs:455]: We still get only one sample ever 60s though, right? This interval places a bound on how stale the sample value can be. 5 seconds seems reasonable to me.

**Comment by erikgrinaker** [comment]
> I was a bit surprised to see that the queue depth metrics tracks messages and not bytes. Do you have thoughts on this?

Mentioned it in https://github.com/neondatabase/neon/pull/9450#issuecomment-2422215540:

> Might be useful to have the byte size of buffered messages -- in aggregate we can easily get this by having counters for WAL bytes received and WAL bytes processed, and compute the difference to find the buffer volume. If we want it per queue we'll need to wrap the mpsc queue to track sizes on send() and recv().

I figured the total buffer volume was probably sufficient -- per queue I think we'd rather track when they get close to blocking, which is a function of message count not size. But let me know if you think we should add a per-queue histogram for it as well.

File: safekeeper/src/receive_wal.rs (hunk 4 above)
Comment (VladLazar) [reply safekeeper/src/receive_wal.rs:455]: Right, they're stored in mem

**Comment by erikgrinaker** [comment]
I realized that we'll have to account for dropped messages in the queue on disconnect. This means we'll need per-queue tracking anyway, so I guess we may as well add a histogram for byte size too.

**Comment by erikgrinaker** [comment]
> I realized that with aggregate buffer volume we'll have to account for dropped messages in the queue on disconnect, otherwise we'll have a metrics leak. This means we'll need per-queue tracking anyway, so I guess we may as well add a histogram for byte size too.

Nevermind, can do this in a `WalAcceptor` drop handler. Skipped the histogram for now, to avoid having to write an MPSC queue wrapper.

Added `safekeeper_wal_receiver_queue_size_total` and `safekeeper_wal_receiver_queue_depth_total`, PTAL.

**Review by VladLazar — APPROVED** [review]

**Pull request merged**